### DISCUSSION TITLE
Linux Desktop Application - How to access visual tree programatically

### DISCUSSION THREAD

**Discussion by HowardPWeiss** [unoplatform/uno#16777] [Q&A]
This is a very basic question

I would like to 

1) Enumerate all the controls in the current canvas programatically
2) Create a new control (TextBlock) at a specific location in the canvas programatically

**Comment by francoistanguay** [comment]
Can you please share an example of what you tried or think should work?

**Comment by HowardPWeiss** [comment]
In XAML, I have an ItemsControl which is bound to a collection of TextBlockObjects

I have defined a TextBlockObject which has properties such as Height, Width, Name, Text, Foreground, Background.  These TextBlockObjects are NOT derived from Microsoft.UI.Xaml.Controls.TextBlockControl

In the method which constructs the MainViewModel, I create TextBlockObjects for each TextBlock I want to display and add them to the collection of TextBlockObjects

This works.  

However, I later want to change the Foreground or Background color of the text block object from a Background task.  In the previous version of Uno (1.75) and .Net 7, this worked.  The foreground/background color changed 

With the latest version of Uno (1.89) and .Net 8., the Foreground/Background does not change.  I have determined that Uno creates a shadow object of the type Microsoft.UI.Xaml.Controls.TextBlockControl for each of the TextBlock objects I add to the ItemsCollection.  When I change the Foreground or Background and I am in the UnoEventLoop Thread, the Foreground/Background changes. When I change the Foreground/Background from a Background thread, the Properties change in the TextBlockObject but do not change in the shadow object

I do not understand why this worked before and does not work now

As a workaround, I created a Microsoft.UI,Xaml.Controls.TextBoxControl as a property each TextBlockObject.  I rewrote the code such that the code now updates the appropriate property in the Microsoft.UI.Xaml.Controls.TextBlockControl.  When I try to change the property from a background thread, I encounter a cannot access property from background thread error.  I therefore changed the code to test if the thread had access to the object and, if not, used a Dispatcher to change the property from the UI thread. 

The problem is that I cannot position the TextBlockObject in the Canvas, In the previous version of Uno, whatever code created the shadow object also set Canvas,Left and Canvas,Top  properties such that the TextBlock appeared in the correct space.

How can I do this with Uno 1.89 and .Net 8?

**Reply by francoistanguay** [reply]
Few pointers: Not sure what version part of Uno you're referring to. Uno is now at version 5.2 (Uno.WInUI package).

Next, in order to change values in your presentation layer so they are applied in your View layer, you need to raise PropertyChanged event on the UI thread using the Dispatcher. Are you using a MVVM framework like CommunityToolkit.MVVM?

If you can please attach a sample with the repro of what isnt working, would be easier for us to investigate and point you in the right direction

**Comment by HowardPWeiss** [comment]
Pardon my confusion

I am developing on Ubuntu 20.04
Ubuntu version 5.2
Visual Studio Code 1.89
dotnet version 8.0.204

**Reply by francoistanguay** [reply]
Can you attach a repro sample with the code that isnt working or an example of what you want to accomplish please?

**Comment by lindexi** [comment]
@HowardPWeiss I create a project named UnoSpySnoop. I think it may help you. See https://github.com/dotnet-campus/UnoSpySnoop

And now I will update it to use the multi-window that host in the same process.
